src/parameterizations/lateral/MOM_mixed_layer_restrat.F90

```diff
@@ -153,35 +166,48 @@ subroutine mixedlayer_restrat_general(h, uhtr, vhtr, tv, fluxes, dt, G, CS)
-      dKm1(:) = dK(:) ! Depth of center of layer K-1
-      dK(:) = dK(:) + 0.5 * ( h(:,j,k) + h(:,j,k-1) ) * G%H_to_m ! Depth of center of layer K
-
-      ! Mixed-layer depth, using sigma-0 (surface reference pressure)
-      deltaRhoAtKm1(:) = deltaRhoAtK(:) ! Store value from previous iteration of K
-      call calculate_density(tv%T(:,j,k), tv%S(:,j,k), pRef_MLD, deltaRhoAtK, is-1, ie-is+3, tv%eqn_of_state)
-      deltaRhoAtK(:) = deltaRhoAtK(:) - rhoSurf(:) ! Density difference between layer K and surface
+  if (CS%MLE_density_diff > 0.) then ! We need to calculate a mixed layer depth, MLD.
+    !! TODO: use derivatives and mid-MLD pressure. Currently this is sigma-0. -AJA
+    pRef_MLD(:) = 0.
+    do j = js-1, je+1
+      dK(:) = 0.5 * h(:,j,1) * G%H_to_m ! Depth of center of surface layer
+      call calculate_density(tv%T(:,j,1), tv%S(:,j,1), pRef_MLD, rhoSurf, is-1, ie-is+3, tv%eqn_of_state)
+      deltaRhoAtK(:) = 0.
+      CS%MLD(:,j) = 0.
+      do k = 2, nz
+        dKm1(:) = dK(:) ! Depth of center of layer K-1
+        dK(:) = dK(:) + 0.5 * ( h(:,j,k) + h(:,j,k-1) ) * G%H_to_m ! Depth of center of layer K
+        ! Mixed-layer depth, using sigma-0 (surface reference pressure)
+        deltaRhoAtKm1(:) = deltaRhoAtK(:) ! Store value from previous iteration of K
+        call calculate_density(tv%T(:,j,k), tv%S(:,j,k), pRef_MLD, deltaRhoAtK, is-1, ie-is+3, tv%eqn_of_state)
+        deltaRhoAtK(:) = deltaRhoAtK(:) - rhoSurf(:) ! Density difference between layer K and surface
+        do i = is-1, ie+1
+          ddRho = deltaRhoAtK(i) - deltaRhoAtKm1(i)
+          if ((CS%MLD(i,j)==0.) .and. (ddRho>0.) .and. &
+              (deltaRhoAtKm1(i)<CS%MLE_density_diff) .and. (deltaRhoAtK(i)>=CS%MLE_density_diff)) then
+            aFac = ( CS%MLE_density_diff - deltaRhoAtKm1(i) ) / ddRho
+            CS%MLD(i,j) = dK(i) * aFac + dKm1(i) * (1. - aFac)
+          endif
+        enddo ! i-loop
+      enddo ! k-loop
       do i = is-1, ie+1
-        ddRho = deltaRhoAtK(i) - deltaRhoAtKm1(i)
-        if ((MLD(i,j)==0.) .and. (ddRho>0.) .and. &
-            (deltaRhoAtKm1(i)<CS%MLE_density_diff) .and. (deltaRhoAtK(i)>=CS%MLE_density_diff)) then
-          aFac = ( CS%MLE_density_diff - deltaRhoAtKm1(i) ) / ddRho
-          MLD(i,j) = dK(i) * aFac + dKm1(i) * (1. - aFac)
-        endif
-      enddo ! i-loop
-
-    enddo ! k-loop
-    do i = is-1, ie+1
-      if ((MLD(i,j)==0.) .and. (deltaRhoAtK(i)<CS%MLE_density_diff)) MLD(i,j) = dK(i) ! Assume mixing to the bottom
-    enddo
-  enddo ! j-loop
+        if ((CS%MLD(i,j)==0.) .and. (deltaRhoAtK(i)<CS%MLE_density_diff)) CS%MLD(i,j) = dK(i) ! Assume mixing to the bottom
+      enddo
+    enddo ! j-loop
+  elseif (CS%MLE_use_PBL_MLD .and. CS%MLE_MLD_decay_time>=0.) then
+    if (.not. associated(MLD)) call MOM_error(FATAL, "MOM_mixedlayer_restrat: "// &
+         "Argument MLD was not associated!")
+    aFac = CS%MLE_MLD_decay_time / ( dt + CS%MLE_MLD_decay_time )
+    bFac = 1. - aFac
```

Comment:
This would be less prone to roundoff if it were written as a ratio of positive definite values as
bFac = dt / (dt + CS%MLE_MLD_decay)